controller/controller_test.go

```diff
@@ -224,6 +224,107 @@ func TestRunOnce(t *testing.T) {
 	assert.Equal(t, math.Float64bits(1), valueFromMetric(verifiedAAAARecords))
 }
 
+// TestRun tests that Run correctly starts and stops
+func TestRun(t *testing.T) {
+	// Fake some desired endpoints coming from our source.
+	source := new(testutils.MockSource)
+	cfg := externaldns.NewConfig()
+	cfg.ManagedDNSRecordTypes = []string{endpoint.RecordTypeA, endpoint.RecordTypeAAAA, endpoint.RecordTypeCNAME}
+	source.On("Endpoints").Return([]*endpoint.Endpoint{
```

Comment:
Possibly extract the setup and validation code that is common between this and TestRunOnce?

Reply:
done